californium-core/src/main/java/org/eclipse/californium/core/observe/ObservationUtil.java

```diff
@@ -48,6 +52,12 @@ public static Observation shallowClone(Observation observation) {
 		clonedRequest.setOptions(request.getOptions());
 		clonedRequest.setPayload(request.getPayload());
 		clonedRequest.setUserContext(request.getUserContext());
+		clonedRequest.addMessageObserver(new MessageObserverAdapter() {
+			@Override
+			public void onResponse(Response response) {
+				request.setResponse(response);
+			}
+		});
```

Comment:
Why do we need this ?
This will make we will have different behavior between a persistent store and the InMemoryObservationStore.

Reply:
> With the introduction of the "clustered notify", unfortunately the behaviour of
Request.waitForResponse() was also changed, but the consequences of that change have never been cleaned up in the code and documentation. It would have been great, if that change to "clustered notify" would have been worked out with more details also for the "plain coap" usage.

I take this for me. This is surely deserved :( ... Introduce this new "clustered notifiy" was not easy and have huge impact on the code + very hard to explain. We expected it could be long  that's why we do that in a 2.0.0 experimental branch. When you start to involve in the project you maybe expect that this part of the code was finished but that was not the case. We always been a bit sceptical about  CoAPClient API and we don't use it, so it's harder to us to detect impact. We even start discussion about the need to have 2 API in californium.
But if you found issues about that and you thought that had to be fixed by us, you should have to let us know.  I'm not sure if this is the sense of you remarks :/

> The CoapClient seems to get downside of this implementation (potentially inefficient large number of NotificationListener) and never benefits from clustering, because clustering would break too much of the API of the CoapClient, or do you see that different?

I agree.  (about NotificationListener I'm not sure this is a real problem, do you experiment issue about that?)

> So there are different ways to go:

Why I do not like the idea to have 2 different behavior ? because that's means tests will check only one behavior.

The initial idea was to go with a waitForResponse which only return response (not notification). We forgot to change the javadoc. CoapClient was changed (with NotificationListener) with that idea in mind.

Reply:
> I agree. (about NotificationListener I'm not sure this is a real problem, do you experiment issue about that?)

I guess this depends just on the numbers of `NotificationListener` which are registered at a CoapEndpoint.

Reply:
Any opinion about just changing the javadoc ?

Reply:
Sure, if the notifies are not longer supported, then the javadoc should be changed.